lib/rexml/parsers/baseparser.rb

```diff
@@ -198,15 +200,10 @@ def pull_event
         #STDERR.puts @source.encoding
         #STDERR.puts "BUFFER = #{@source.buffer.inspect}"
         if @document_status == nil
-          word = @source.match( /\A((?:\s+)|(?:<[^>]*>))/um )
-          word = word[1] unless word.nil?
-          #STDERR.puts "WORD = #{word.inspect}"
-          case word
-          when COMMENT_START
-            return [ :comment, @source.match( COMMENT_PATTERN, true )[1] ]
-          when XMLDECL_START
+          @source.read
```

Comment:
In general, the following pseudo code is a common `StringScanner` usage:

```ruby
@scanner.scan(/\s*/) # skip spaces
if @scanner.scan("<!")
  if @scanner.scan("--")
    # parse comment
  elsif @scanner.scan(/DOCTYPE\s/)
    # parse DOCTYPE
  else
    raise "invalid XML"
  end
elsif @scanner.scan("<?")
  if @scanner.scan(/xml\s/)
    # parse XML declaration
  else
    # parse processing instruction
  end
else
  document_status = :after_doctype
end
```

Could you use this style? We don't need to use existing regular expression constants such as `DOCTYPE_START` for it. Because match unit will be different for the current approach and `StringScanner` based approach.